vault_cert_revoker/main.go

```diff
@@ -0,0 +1,240 @@
+func QueryVictoriaMetrics(config *Config) ([]Certificate, error) {
+	client, err := CreateTLSClient(config)
+	if err != nil {
+		slog.Error("Failed to create TLS client", "error", err)
+		return nil, err
+	}
+
+	queryParams := url.Values{}
+	queryParams.Set("query", config.VMQuery)
+	queryURL := fmt.Sprintf("%s?%s", config.VMURL, queryParams.Encode())
+
+	resp, err := client.Get(queryURL)
+	if err != nil {
+		slog.Error("Failed to query VictoriaMetrics", "error", err)
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		slog.Error("VictoriaMetrics query failed", "status", resp.Status)
+		return nil, fmt.Errorf("VM query failed: %s", resp.Status)
+	}
+
+	var result struct {
+		Data struct {
+			Result []struct {
+				Metric map[string]string `json:"metric"`
+				Value  []interface{}     `json:"value"`
+			} `json:"result"`
+		} `json:"data"`
+	}
+
+	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
+		slog.Error("Failed to decode VictoriaMetrics response", "error", err)
+		return nil, err
+	}
+
+	var certs []Certificate
+	for _, res := range result.Data.Result {
+		certs = append(certs, Certificate{
+			CommonName:         res.Metric["common_name"],
+			OrganizationalUnit: res.Metric["organizational_unit"],
+			SerialNumber:       res.Metric["serial"],
+		})
+	}
+
+	slog.Info("Certificates retrieved", "count", len(certs))
+	return certs, nil
+}
+
+// FilterCertificates filters certificates by common_name using a regex
+// TODO - might be nice to filter OU too?
+func FilterCertificates(certs []Certificate, regex *regexp.Regexp) []Certificate {
+	var filtered []Certificate
+	for _, cert := range certs {
+		if regex.MatchString(cert.CommonName) {
+			filtered = append(filtered, cert)
+		}
+	}
+	return filtered
+}
+
+// RevokeCertificate uses the Vault CLI to revoke a certificate
+// Note - can't revoke expired certs
+// https://github.com/hashicorp/vault/issues/19452
+func RevokeCertificate(config *Config, cert Certificate, dryRun bool) error {
+	vaultCommand := "vault"
+	args := []string{
+		"write",
+		fmt.Sprintf("%s/revoke", config.VaultPKIPath),
+		fmt.Sprintf("serial_number=%s", cert.SerialNumber),
+	}
+
+	if dryRun {
+		slog.Info("[DRY RUN] Would revoke certificate", "serial_number", cert.SerialNumber, "common_name", cert.CommonName)
+		return nil
+	}
+
+	slog.Info("Revoking certificate using Vault CLI", "serial_number", cert.SerialNumber, "command", vaultCommand, "args", args)
+
+	cmd := exec.Command(vaultCommand, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		slog.Error("Vault CLI command failed", "error", err)
+		return err
+	}
+
+	slog.Info("Certificate revoked successfully", "serial_number", cert.SerialNumber)
+	return nil
+}
+
+func Confirm(prompt string) bool {
+	fmt.Print(prompt + " [y/N]: ")
+	scanner := bufio.NewScanner(os.Stdin)
+	if scanner.Scan() {
+		answer := scanner.Text()
+		return answer == "y" || answer == "Y"
```

Comment:
My preferred way for naive evaluation of stuff like this is something like `answer.lower()[0] == "y"`

Reply:
makes sense too, can't think of any downsides with my lamer method though